azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -650,7 +624,7 @@ def on_timeout():
             logger.debug("{}({}): Sending down".format(self.name, op.name))
             self.send_op_down(op)
         else:
-            self.send_op_down(op)
+            super(OpTimeoutStage, self)._run_op(op)
 
```

Comment:
I'm still chewing on this change and i'm not sure if making PipelineStage._run_op concrete was a good call or not.
This changes the fundamental nature of _run_op from "do everything necessary to handle and pass on" to "do a few special things and let the base class handle everything else".  This specific code illustrates one of my concerns.  Why does this line use super._run_op(), but the code 3 lines up uses send_op_down?  The behavior is the same (right now), but we've introduced a new convention for sending ops to the next stage.  Even though there is no functional distinction between the 2 right now, there is a conceptual distinction, and a functional distinction will eventually develop.

Reply:
I'm open to changing this back. This made more sense before we stopped sharing the test for this conditional branch.

---
In reply to: [354859717](https://github.com/Azure/azure-iot-sdk-python/pull/409#discussion_r354859717) [](ancestors = 354859717)